flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsMergingCheckpointStorageLocation.java

```diff
@@ -80,6 +80,9 @@ public FsMergingCheckpointStorageLocation(
                                 reference,
                                 fileStateSizeThreshold,
                                 writeBufferSize);
+
+        // notify file merging snapshot manager when FsMergingCheckpointStorageLocation create.
+        fileMergingSnapshotManager.notifyCheckpointStart(subtaskKey, checkpointId);
```

Comment:
I'd suggest move this part to `SubtaskCheckpointCoordinatorImpl#checkpointState` (before `initInputsCheckpoint`, which may resolve the location), since the location could be created twice for one checkpoint.

Reply:
Good suggestion, done.